### ISSUE TITLE
Windows: :edit filename inconsistencies

### ISSUE THREAD

**Issue by equalsraf** [neovim/neovim#3912]
E.g. try the following

    :e \s3go.
    :w

- Vim/Windows - fails on `:write` with `E212: Can't open file for writing`
- Vim/Unix - succeeds, names the buffer on edit as `s3go.` and creates the file `s3go.` instead of `\s3go.`
- Neovim/Unix behaves like Vim/Unix
- Neovim/Windows behaves like Vim/Unix too (#810)

What is the desired behaviour here?

**Comment by ZyX-I** [comment]
This is the direct consequence of backslash used both for path separator and as escape character. Since `/` is a valid path separator on Windows to I would personally prefer to force backslash escaping to get a backslash: so that path `C:\Program Files\foo` must be spelled as either `C:\\Program\ Files\\foo` or as `C:/Program\ Files/foo` for use with `:edit`. Do not know what exactly Neovim is doing, but Vim treats backslash depending on the next character: if it is something that can be escaped (but **not** something like `*`: you can’t escape asterisk on Windows AFAIR) then it is escape character, if it something else then it is path separator. As I said (about asterisk) set of characters that need escaping is not full (AFAIR I did report this to vim-dev, but that never got fixed), and such behaviour is incosistent.

Since just changing \\ paths to / paths and using `fnameescape()` for non-hardcoded paths is all that needs to be done to be done to be compatible with both Vim and Neovim and most plugins are either \*nix-only or cross-platform (and thus use `/` or things like `os.path.join`) I do not think forcing backslash to be an escape character always and treating it as a path separator only in unescaped path will do much harm.

**Comment by justinmk** [comment]
gvim.exe and vim.exe on Windows follow some weird rules here, though it is _usually_ convenient. It will surprise the user if they use the same vimrc for gvim.exe, cygwin/MSYS2 Vim.

>  Vim treats backslash depending on the next character: if it is something that can be escaped (but not something like *: you can’t escape asterisk on Windows AFAIR) then it is escape character, if it something else then it is path separator. As I said (about asterisk) set of characters that need escaping is not full (AFAIR I did report this to vim-dev, but that never got fixed), and such behaviour is incosistent.

Yes, I've encountered this paradox with [vim-dirvish](https://github.com/justinmk/vim-dirvish/blob/master/autoload/dirvish.vim#L377): I use `expand(fnameescape()` to get a normalized path, but that breaks `:edit` if the path contains a literal `#` or `%`. Not pleasant for plugin authors.

> Since just changing \ paths to / paths and using fnameescape() for non-hardcoded paths is all that needs to be done to be done to be compatible with both Vim and Neovim and most plugins are either *nix-only or cross-platform (and thus use / or things like os.path.join) I do not think forcing backslash to be an escape character always and treating it as a path separator only in unescaped path will do much harm.

:+1: Because this is a least-effort approach, let's try this and see if anyone cares that `:edit` is slightly less convenient.

**Comment by ZyX-I** [comment]
@justinmk I was wrong about asterisk: I now remember the bug and it was about escaping `$`, exactly for `expand(fnameescape())` normalization: should path contain `$` you can’t use this hack because `$var` gets expanded. I do not remember whether or not I actually had any problems with asterisk, but this is also possible. Also can’t remember now whether it was bug on `fnameescape()` stage only (`$` does not get escaped) or both at `fnameescape()` and `expand()` stages (`\$` is treated as `/$`). Will check in wine now.

Apparently you also saw this or similar problem with `%`.

**Comment by ZyX-I** [comment]
On Windows:

1. `fnameescape('$PATH') is# '$PATH'`. Not good: `$` not escaped.
2. `expand('foo$PATH') is# 'fooC:\…'`. Not good: valid path turned into a junk.
3. `expand('foo\$PATH') is# 'foo\$PATH'`. Not good: cannot use escaping to make valid path not expand. (To actually edit e.g. `C:\foo$PATH` one needs to use <code>:edit `="C:/foo$PATH"`</code>. But `fnameescape()` is not helpful and <code>`=</code> cannot be used on linux due to problems with valid paths with newlines.)
4. `fnameescape('foo*')` is# `foo\*`. Good? No:
5. `expand('foo\*') is# ''`. Not good: you can’t escape asterisk.
6. `fnameescape('foo%') is# 'foo\%'`. Good? I guess this has the same problem.

**Comment by ZyX-I** [comment]
This is wine 32-bit XP emulation 1.7.55, vim without cream gvim.exe 7.4.977.

**Comment by ZyX-I** [comment]
Basically it is not only that you can’t really escape anything except space, you *must not* use `fnameescape()` on Windows.

**Comment by justinmk** [comment]
Confirmed your results on windows 7, gvim 7.4.827.

> you must not use fnameescape() on Windows.

That rule won't be necessary for Neovim though.

**Comment by ZyX-I** [comment]
Actually `%` and `#` *can* be used here: `:edit foo\%` edits file `foo%` (though note that `edit foo%` when `@%` is `foo$PATH` results in editing `fooC:\…`).

**Comment by ZyX-I** [comment]
But `expand()` is a problem: it does not remove backslash.

**Comment by justinmk** [comment]
> Actually % and # can be used here

Yes, I mis-remembered the issue. The actual problem is that `isdirectory('C:\Users\jkeyes\foo\#bar\')` returns `0`. So in order to check if the directory exists, the normalized path must now be un-escaped, which is basically impossible.

**Comment by ZyX-I** [comment]
@justinmk Note the linked vim/vim#541: I have some more details there (specifically regarding backticks).

**Comment by ZyX-I** [comment]
After this investigation I would say that backwards compatibility *does not deserve* to be kept here. It is not necessary for my approach to be used, but any used approach should allow the following:

1. `execute 'edit' fnameescape(fname)` should open buffer whose name is *exactly* the string stored in `fname`, no matter what its contents is.
2. `expand(fnameescape(fname))` should yield `fname` for every possible `fname` that is the result of the evaluation of `expand(fnameescape(base_fname))`.
3. `expand(fnameescape(fname))` should return exactly one file name for every possible `fname` string.
4. `glob(fnameescape(fname) . '/*', 1, 1)` should list all (except for usual dot problems) files in directory `fname`, no matter what its name is as long as it is the name of the existing directory.
